sway-lib-std/src/bytes.sw

```diff
@@ -0,0 +1,849 @@
+    /// vec.push(b);
+    /// vec.push(c);
+    ///
+    /// let bytes = Bytes::from_vec_u8(vec);
+    ///
+    /// assert(bytes.len == 3);
+    /// assert(bytes.get(0).unwrap() == a);
+    /// assert(bytes.get(1).unwrap() == b);
+    /// assert(bytes.get(2).unwrap() == c);
+    /// ```
+    pub fn from_vec_u8(ref mut vec: Vec<u8>) -> Self {
+        let mut bytes = Bytes::new();
+        let mut i = 0;
+        let length = vec.len();
+        while i < length {
+            bytes.push(vec.get(i).unwrap());
+            i += 1;
+        };
+        bytes
+    }
+
+    /// Creates a Vec<u8> from a Bytes.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// use std:bytes::Bytes;
+    ///
+    /// let mut bytes = Bytes::new();
+    /// let a = 5u8;
+    /// let b = 7u8;
+    /// let c = 9u8
+    /// bytes.push(a);
+    /// bytes.push(b);
+    /// bytes.push(c);
+    ///
+    /// assert(bytes.len() == 3);
+    ///
+    /// let vec = bytes.into_vec_u8();
+    ///
+    /// assert(vec.len() == 3);
+    /// assert(vec.get(0).unwrap() == a);
+    /// assert(vec.get(1).unwrap() == b);
+    /// assert(vec.get(2).unwrap() == c);
+    /// ```
+    pub fn into_vec_u8(self) -> Vec<u8> {
+        let mut vec = Vec::new();
+        let mut i = 0;
+        let length = self.len;
+        while i < length {
+            vec.push(self.get(i).unwrap());
+            i += 1;
+        };
+        vec
+    }
+
+    pub fn split(self, index: u64) -> (Bytes, Bytes) {
+        // TODO: handle edge cases:
+        // index == 0 (also handles the self.len() == 1 case)
+        // index == self.len() - maybe we say index must be < self.len - 1
+        // @todo check rust impl for this type of thing
+        assert(index < self.len());
+        let mut first = Bytes::with_capacity(index);
+        let mut second = Bytes::with_capacity(self.len() - index);
+        let mut i = 0;
+        while i < index {
+            first.push(self.get(i).unwrap());
+            i += 1;
+        };
+
+        let mut i2 = index;
+        while i2 < self.len {
+            second.push(self.get(i2).unwrap());
+            i2 += 1;
+        };
+        (first, second)
+    }
+
+    pub fn join(ref mut self, other: self) -> Self {
+        let mut joined = Bytes::with_capacity(self.len + other.len);
+
+        let mut i = 0;
+        while i < self.len() {
+            joined.push(self.get(i).unwrap());
+            i += 1;
+        };
+
+        let mut i2 = 0;
+        while i2 < other.len() {
+            joined.push(other.get(i2).unwrap());
+            i2 += 1;
+        };
+
+        joined
+    }
+}
+
+impl core::ops::Eq for Bytes {
+    fn eq(self, other: Self) -> bool {
+        if self.len() == other.len() {
+            let mut i = self.len();
+            while i > 0 {
+                assert(self.get(i - 1).unwrap() == other.get(i - 1).unwrap());
+                i -= 1;
+            };
+            true
+        } else {
+            false
+        }
```

Comment:
You may want to exit early instead of `if/else` to reduce indentation for the `while` above. So:
```rust
if self.len() != other.len() {
    return false;
}

while ...
```